File: SchoolWebApp-API/Project.Core/Entities/Finance/Budget.cs
using SchoolWebApp.Core.Entities.Academics;
using SchoolWebApp.Core.Entities.School;
using SchoolWebApp.Core.Entities.Shared;
using System.ComponentModel.DataAnnotations;

namespace SchoolWebApp.Core.Entities.Finance
{
    public class Budget : Base
    {
        [Required]
        [StringLength(255)]
        public required string Name { get; set; }

        [StringLength(500)]
        public string? Description { get; set; }

        public DateTime StartDate { get; set; }
        public DateTime EndDate { get; set; }

        public int AcademicYearId { get; set; }
        public AcademicYear? AcademicYear { get; set; }

        public int DepartmentId { get; set; }
        public Department? Department { get; set; }

        public int? BudgetMasterId { get; set; }
        public BudgetMaster? BudgetMaster { get; set; }

        public bool IsActive { get; set; } = true;

        public List<BudgetLine> Lines { get; set; } = new();
        public List<BudgetAmendment> Amendments { get; set; } = new();
    }
}

File: SchoolWebApp-Client/src/app/finance/models/budget.ts
import {ResourceModel} from '@/core/models/ResourceModel';

export class BudgetLine extends ResourceModel<BudgetLine> {
    public budgetId?: number;
    public accountId?: number;
    public accountCode?: string;
    public accountName?: string;
    public accountType?: string;
    public budgetedAmount?: number;
    public amendedAmount?: number;
    public effectiveAmount?: number;
    public actualAmount?: number;
    public variance?: number;
    public notes?: string;

    constructor(model?: Partial<BudgetLine>) {
        super(model);
    }
}

export class Budget extends ResourceModel<Budget> {
    public name?: string;
    public description?: string;
    public startDate?: string;
    public endDate?: string;
    public academicYearId?: number;
    public academicYearName?: string;
    public departmentId?: number;
    public departmentName?: string;
    public budgetMasterId?: number;
    public budgetMasterName?: string;
    public isActive?: boolean;
    public totalBudgeted?: number;
    public totalAmendments?: number;
    public totalEffective?: number;
    public lines!: BudgetLine[];

    constructor(model?: Partial<Budget>) {
        super(model);
        if (!this.lines) this.lines = [];
    }
}
